src/keymgmt.c

```diff
@@ -482,14 +482,19 @@ static int p11prov_common_gen(struct key_generator *ctx,
     }
 
     ret = p11prov_merge_pub_attrs_into_priv(pub_key, priv_key);
+    if (ret != CKR_OK) {
+        goto done;
+    }
+
+    ret = p11prov_set_pub(priv_key, pub_key);
 
 done:
     if (ret != CKR_OK) {
+        p11prov_obj_free(pub_key);
```

Comment:
This is wrong now because the next line will already free the pub_key you set on it.
After you set the pub key on the priv key you need to set pub_key = NULL;

This is may be one of the segfault issue, because you are causing a double-free here.